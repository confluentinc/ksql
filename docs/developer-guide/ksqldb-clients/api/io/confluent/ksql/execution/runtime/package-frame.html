<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_265) on Fri Oct 01 11:59:34 CDT 2021 -->
<title>io.confluent.ksql.execution.runtime</title>
<meta name="date" content="2021-10-01">
<link rel="stylesheet" type="text/css" href="../../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../../script.js"></script>
</head>
<body>
<h1 class="bar"><a href="../../../../../io/confluent/ksql/execution/runtime/package-summary.html" target="classFrame">io.confluent.ksql.execution.runtime</a></h1>
<div class="indexContainer">
<h2 title="Classes">Classes</h2>
<ul title="Classes">
<li><a href="RuntimeBuildContext.html" title="class in io.confluent.ksql.execution.runtime" target="classFrame">RuntimeBuildContext</a></li>
<li><a href="RuntimeBuildContextTest.html" title="class in io.confluent.ksql.execution.runtime" target="classFrame">RuntimeBuildContextTest</a></li>
</ul>
</div>
</body>
</html>
