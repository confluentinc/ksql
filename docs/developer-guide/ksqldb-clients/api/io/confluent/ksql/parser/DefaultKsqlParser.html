<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_265) on Fri Oct 01 11:59:30 CDT 2021 -->
<title>DefaultKsqlParser</title>
<meta name="date" content="2021-10-01">
<link rel="stylesheet" type="text/css" href="../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="DefaultKsqlParser";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":9,"i1":10,"i2":9,"i3":10};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../io/confluent/ksql/parser/ColumnReferenceParserTest.html" title="class in io.confluent.ksql.parser"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../../../io/confluent/ksql/parser/DefaultTraversalVisitor.html" title="class in io.confluent.ksql.parser"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../index.html?io/confluent/ksql/parser/DefaultKsqlParser.html" target="_top">Frames</a></li>
<li><a href="DefaultKsqlParser.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">io.confluent.ksql.parser</div>
<h2 title="Class DefaultKsqlParser" class="title">Class DefaultKsqlParser</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>io.confluent.ksql.parser.DefaultKsqlParser</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd><a href="../../../../io/confluent/ksql/parser/KsqlParser.html" title="interface in io.confluent.ksql.parser">KsqlParser</a></dd>
</dl>
<hr>
<br>
<pre>public class <span class="typeNameLabel">DefaultKsqlParser</span>
extends java.lang.Object
implements <a href="../../../../io/confluent/ksql/parser/KsqlParser.html" title="interface in io.confluent.ksql.parser">KsqlParser</a></pre>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== NESTED CLASS SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="nested.class.summary">
<!--   -->
</a>
<h3>Nested Class Summary</h3>
<ul class="blockList">
<li class="blockList"><a name="nested.classes.inherited.from.class.io.confluent.ksql.parser.KsqlParser">
<!--   -->
</a>
<h3>Nested classes/interfaces inherited from interface&nbsp;io.confluent.ksql.parser.<a href="../../../../io/confluent/ksql/parser/KsqlParser.html" title="interface in io.confluent.ksql.parser">KsqlParser</a></h3>
<code><a href="../../../../io/confluent/ksql/parser/KsqlParser.ParsedStatement.html" title="class in io.confluent.ksql.parser">KsqlParser.ParsedStatement</a>, <a href="../../../../io/confluent/ksql/parser/KsqlParser.PreparedStatement.html" title="class in io.confluent.ksql.parser">KsqlParser.PreparedStatement</a>&lt;<a href="../../../../io/confluent/ksql/parser/KsqlParser.PreparedStatement.html" title="type parameter in KsqlParser.PreparedStatement">T</a> extends <a href="../../../../io/confluent/ksql/parser/tree/Statement.html" title="class in io.confluent.ksql.parser.tree">Statement</a>&gt;</code></li>
</ul>
</li>
</ul>
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Field Summary table, listing fields, and an explanation">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Field and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static org.antlr.v4.runtime.BaseErrorListener</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../io/confluent/ksql/parser/DefaultKsqlParser.html#ERROR_VALIDATOR">ERROR_VALIDATOR</a></span></code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../io/confluent/ksql/parser/DefaultKsqlParser.html#DefaultKsqlParser--">DefaultKsqlParser</a></span>()</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t1" class="tableTab"><span><a href="javascript:show(1);">Static Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>static io.confluent.ksql.parser.SqlBaseParser.StatementsContext</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../io/confluent/ksql/parser/DefaultKsqlParser.html#getParseTree-java.lang.String-">getParseTree</a></span>(java.lang.String&nbsp;sql)</code>&nbsp;</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>java.util.List&lt;<a href="../../../../io/confluent/ksql/parser/KsqlParser.ParsedStatement.html" title="class in io.confluent.ksql.parser">KsqlParser.ParsedStatement</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../io/confluent/ksql/parser/DefaultKsqlParser.html#parse-java.lang.String-">parse</a></span>(java.lang.String&nbsp;sql)</code>
<div class="block">Parse the supplied <code>sql</code> into a list of statements.</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>static <a href="../../../../io/confluent/ksql/parser/KsqlParser.ParsedStatement.html" title="class in io.confluent.ksql.parser">KsqlParser.ParsedStatement</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../io/confluent/ksql/parser/DefaultKsqlParser.html#parsedStatement-io.confluent.ksql.parser.SqlBaseParser.SingleStatementContext-">parsedStatement</a></span>(io.confluent.ksql.parser.SqlBaseParser.SingleStatementContext&nbsp;statement)</code>&nbsp;</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code><a href="../../../../io/confluent/ksql/parser/KsqlParser.PreparedStatement.html" title="class in io.confluent.ksql.parser">KsqlParser.PreparedStatement</a>&lt;?&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../io/confluent/ksql/parser/DefaultKsqlParser.html#prepare-io.confluent.ksql.parser.KsqlParser.ParsedStatement-io.confluent.ksql.metastore.TypeRegistry-">prepare</a></span>(<a href="../../../../io/confluent/ksql/parser/KsqlParser.ParsedStatement.html" title="class in io.confluent.ksql.parser">KsqlParser.ParsedStatement</a>&nbsp;stmt,
       <a href="../../../../io/confluent/ksql/metastore/TypeRegistry.html" title="interface in io.confluent.ksql.metastore">TypeRegistry</a>&nbsp;typeRegistry)</code>
<div class="block">Prepare the supplied <code>statement</code>.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ FIELD DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a name="ERROR_VALIDATOR">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>ERROR_VALIDATOR</h4>
<pre>public static final&nbsp;org.antlr.v4.runtime.BaseErrorListener ERROR_VALIDATOR</pre>
</li>
</ul>
</li>
</ul>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="DefaultKsqlParser--">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>DefaultKsqlParser</h4>
<pre>public&nbsp;DefaultKsqlParser()</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="parse-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>parse</h4>
<pre>public&nbsp;java.util.List&lt;<a href="../../../../io/confluent/ksql/parser/KsqlParser.ParsedStatement.html" title="class in io.confluent.ksql.parser">KsqlParser.ParsedStatement</a>&gt;&nbsp;parse(java.lang.String&nbsp;sql)</pre>
<div class="block"><span class="descfrmTypeLabel">Description copied from interface:&nbsp;<code><a href="../../../../io/confluent/ksql/parser/KsqlParser.html#parse-java.lang.String-">KsqlParser</a></code></span></div>
<div class="block">Parse the supplied <code>sql</code> into a list of statements.</div>
<dl>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code><a href="../../../../io/confluent/ksql/parser/KsqlParser.html#parse-java.lang.String-">parse</a></code>&nbsp;in interface&nbsp;<code><a href="../../../../io/confluent/ksql/parser/KsqlParser.html" title="interface in io.confluent.ksql.parser">KsqlParser</a></code></dd>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>sql</code> - the sql to parse.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the list of parsed statements.</dd>
</dl>
</li>
</ul>
<a name="parsedStatement-io.confluent.ksql.parser.SqlBaseParser.SingleStatementContext-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>parsedStatement</h4>
<pre>public static&nbsp;<a href="../../../../io/confluent/ksql/parser/KsqlParser.ParsedStatement.html" title="class in io.confluent.ksql.parser">KsqlParser.ParsedStatement</a>&nbsp;parsedStatement(io.confluent.ksql.parser.SqlBaseParser.SingleStatementContext&nbsp;statement)</pre>
</li>
</ul>
<a name="prepare-io.confluent.ksql.parser.KsqlParser.ParsedStatement-io.confluent.ksql.metastore.TypeRegistry-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>prepare</h4>
<pre>public&nbsp;<a href="../../../../io/confluent/ksql/parser/KsqlParser.PreparedStatement.html" title="class in io.confluent.ksql.parser">KsqlParser.PreparedStatement</a>&lt;?&gt;&nbsp;prepare(<a href="../../../../io/confluent/ksql/parser/KsqlParser.ParsedStatement.html" title="class in io.confluent.ksql.parser">KsqlParser.ParsedStatement</a>&nbsp;stmt,
                                               <a href="../../../../io/confluent/ksql/metastore/TypeRegistry.html" title="interface in io.confluent.ksql.metastore">TypeRegistry</a>&nbsp;typeRegistry)</pre>
<div class="block"><span class="descfrmTypeLabel">Description copied from interface:&nbsp;<code><a href="../../../../io/confluent/ksql/parser/KsqlParser.html#prepare-io.confluent.ksql.parser.KsqlParser.ParsedStatement-io.confluent.ksql.metastore.TypeRegistry-">KsqlParser</a></code></span></div>
<div class="block">Prepare the supplied <code>statement</code>.</div>
<dl>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code><a href="../../../../io/confluent/ksql/parser/KsqlParser.html#prepare-io.confluent.ksql.parser.KsqlParser.ParsedStatement-io.confluent.ksql.metastore.TypeRegistry-">prepare</a></code>&nbsp;in interface&nbsp;<code><a href="../../../../io/confluent/ksql/parser/KsqlParser.html" title="interface in io.confluent.ksql.parser">KsqlParser</a></code></dd>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>stmt</code> - the statement to build</dd>
<dd><code>typeRegistry</code> - the type registry</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the prepared statement.</dd>
</dl>
</li>
</ul>
<a name="getParseTree-java.lang.String-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>getParseTree</h4>
<pre>public static&nbsp;io.confluent.ksql.parser.SqlBaseParser.StatementsContext&nbsp;getParseTree(java.lang.String&nbsp;sql)</pre>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../io/confluent/ksql/parser/ColumnReferenceParserTest.html" title="class in io.confluent.ksql.parser"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../../../io/confluent/ksql/parser/DefaultTraversalVisitor.html" title="class in io.confluent.ksql.parser"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../index.html?io/confluent/ksql/parser/DefaultKsqlParser.html" target="_top">Frames</a></li>
<li><a href="DefaultKsqlParser.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
