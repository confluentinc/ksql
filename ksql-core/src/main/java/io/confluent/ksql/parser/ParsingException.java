/**
 * Copyright 2017 Confluent Inc.
 **/
package io.confluent.ksql.parser;

import io.confluent.ksql.parser.tree.NodeLocation;
import org.antlr.v4.runtime.RecognitionException;

import static java.lang.String.format;

public class ParsingException
    extends RuntimeException {

  private final int line;
  private final int charPositionInLine;

  public ParsingException(String message, RecognitionException cause, int line, int charPositionInLine) {
    super(message, cause);

    this.line = line;
    this.charPositionInLine = charPositionInLine;
  }

  public ParsingException(String message) {
    this(message, null, 1, 0);
  }

  public ParsingException(String message, NodeLocation nodeLocation) {
    this(message, null, nodeLocation.getLineNumber(), nodeLocation.getColumnNumber());
  }

  public int getLineNumber() {
    return line;
  }

  public int getColumnNumber() {
    return charPositionInLine + 1;
  }

  public String getErrorMessage() {
    return super.getMessage();
  }

  @Override
  public String getMessage() {
    return format("line %s:%s: %s", getLineNumber(), getColumnNumber(), getErrorMessage());
  }
}
