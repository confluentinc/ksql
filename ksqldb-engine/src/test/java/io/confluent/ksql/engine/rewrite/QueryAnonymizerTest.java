/*
 * Copyright 2021 Confluent Inc.
 * Licensed under the Confluent Community License (the "License"); you may not use
 * this file except in compliance with the License.  You may obtain a copy of the
 * License at
 * http://www.confluent.io/confluent-community-license
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations under the License.
 */

package io.confluent.ksql.engine.rewrite;

import io.confluent.ksql.parser.DefaultKsqlParser;
import io.confluent.ksql.parser.ParsingException;
import org.antlr.v4.runtime.ParserRuleContext;
import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

public class QueryAnonymizerTest {

  private QueryAnonymizer anon;

  @Before
  public void setUp() {
    anon = new QueryAnonymizer();
  }

  @Test
  public void shouldWorkAsExpectedWhenPassedAParseTreeInsteadOfString() {
    // Given:
    final ParserRuleContext tree =
        DefaultKsqlParser.getParseTree("DESCRIBE my_stream EXTENDED;");

    // Then:
    Assert.assertEquals("DESCRIBE stream1 EXTENDED;",
        anon.anonymize(tree));
  }

  @Test
  public void shouldThrowWhenUnparseableStringProvided() {
    // Given:
    final String nonsenseUnparsedQuery = "cat";
    // Then:
    Assert.assertThrows(ParsingException.class, () -> anon.anonymize(nonsenseUnparsedQuery));
  }

  @Test
  public void shouldLeaveListAndPrintStatementsAsTheyAre() {
    Assert.assertEquals("LIST PROPERTIES;", anon.anonymize("LIST PROPERTIES;"));
    Assert.assertEquals("SHOW PROPERTIES;", anon.anonymize("SHOW PROPERTIES;"));
    Assert.assertEquals("LIST TOPICS;", anon.anonymize("LIST TOPICS;"));
    Assert.assertEquals("SHOW ALL TOPICS EXTENDED;",
        anon.anonymize("SHOW ALL TOPICS EXTENDED;"));
    Assert.assertEquals("LIST STREAMS EXTENDED;",
        anon.anonymize("LIST STREAMS EXTENDED;"));
    Assert.assertEquals("LIST FUNCTIONS;", anon.anonymize("LIST FUNCTIONS;"));
    Assert.assertEquals("LIST CONNECTORS;", anon.anonymize("LIST CONNECTORS;"));
    Assert.assertEquals("LIST SOURCE CONNECTORS;",
        anon.anonymize("LIST SOURCE CONNECTORS;"));
    Assert.assertEquals("LIST TYPES;", anon.anonymize("LIST TYPES;"));
    Assert.assertEquals("LIST VARIABLES;", anon.anonymize("LIST VARIABLES;"));
    Assert.assertEquals("LIST QUERIES;", anon.anonymize("LIST QUERIES;"));
  }

  @Test
  public void describeStatementsShouldGetAnonymized() {
    Assert.assertEquals("DESCRIBE stream1 EXTENDED;",
        anon.anonymize("DESCRIBE my_stream EXTENDED;"));
    Assert.assertEquals("DESCRIBE STREAMS EXTENDED;",
        anon.anonymize("DESCRIBE STREAMS EXTENDED;"));
    Assert.assertEquals("DESCRIBE FUNCTION function;",
        anon.anonymize("DESCRIBE FUNCTION my_function;"));
    Assert.assertEquals("DESCRIBE CONNECTOR connector;",
        anon.anonymize("DESCRIBE CONNECTOR my_connector;"));
  }

  @Test
  public void printStatementsShouldGetAnonymized() {
    Assert.assertEquals("PRINT topic FROM BEGINNING;",
        anon.anonymize("PRINT my_topic FROM BEGINNING;"));
    Assert.assertEquals("PRINT topic INTERVAL '0';",
        anon.anonymize("PRINT my_topic INTERVAL 2;"));
    Assert.assertEquals("PRINT topic LIMIT '0';",
        anon.anonymize("PRINT my_topic LIMIT 3;"));
  }

  @Test
  public void terminateQueryShouldGetAnonymized() {
    Assert.assertEquals("TERMINATE query;",
        anon.anonymize("TERMINATE my_query;"));
    Assert.assertEquals("TERMINATE ALL;",
        anon.anonymize("TERMINATE ALL;"));
  }

  @Test
  public void shouldAnonymizeSetUnsetProperty() {
    Assert.assertEquals("SET 'auto.offset.reset'='[string]';",
        anon.anonymize("SET 'auto.offset.reset'='earliest';"));
    Assert.assertEquals("UNSET 'auto.offset.reset';",
        anon.anonymize("UNSET 'auto.offset.reset';"));
  }

  @Test
  public void shouldAnonymizeDefineUndefineProperty() {
    Assert.assertEquals("DEFINE variable='[string]';",
        anon.anonymize("DEFINE format = 'JSON';"));
    Assert.assertEquals("UNDEFINE variable;",
        anon.anonymize("UNDEFINE format;"));
  }

  @Test
  public void shouldAnonymizeSelectStatementCorrectly() {
    Assert.assertEquals("SELECT * FROM source1;",
        anon.anonymize("SELECT * FROM S1;"));
  }

  @Test
  public void shouldAnonymizeExplainStatementCorrectly() {
    Assert.assertEquals("EXPLAIN query;", anon.anonymize("EXPLAIN my_query;"));
    Assert.assertEquals("EXPLAIN SELECT * FROM source1;",
        anon.anonymize("EXPLAIN SELECT * from S1;"));
  }

  @Test
  public void shouldAnonymizeJoinStatementsCorrectly() {
    final String output = anon.anonymize("INSERT INTO OUTPUT SELECT col1, col2, col3"
        + " FROM SOURCE1 S1 JOIN SOURCE2 S2 WITHIN 1 SECOND ON col1.k=col2.k;");

    Approvals.verify(output);
  }

  @Test
  public void shouldAnonymizeCreateStreamQueryCorrectly() {
    final String output = anon.anonymize(
        "CREATE STREAM my_stream (profileId VARCHAR, latitude DOUBLE, longitude DOUBLE)\n"
        + "WITH (kafka_topic='locations', value_format='json', partitions=1);");

    Approvals.verify(output);
  }

  @Test
  public void shouldAnonymizeCreateStreamAsQueryCorrectly() {
    final String output = anon.anonymize(
        "CREATE STREAM my_stream AS SELECT user_id, browser_cookie, ip_address\n"
            + "FROM another_stream\n"
            + "WHERE user_id = 4214\n"
            + "AND browser_cookie = 'aefde34ec'\n"
            + "AND ip_address = '10.10.0.2';");

    Approvals.verify(output);
  }

  @Test
  public void shouldAnonymizeCreateTableCorrectly() {
    final String output = anon.anonymize(
        "CREATE TABLE my_table (profileId VARCHAR, latitude DOUBLE, longitude DOUBLE)\n"
            + "WITH (kafka_topic='locations', value_format='json', partitions=1);");

    Approvals.verify(output);
  }

  @Test
  public void shouldAnonymizeCreateTableAsQueryCorrectly() {
    final String output = anon.anonymize(
        "CREATE TABLE my_table AS SELECT user_id, browser_cookie, ip_address\n"
            + "FROM another_table\n"
            + "WHERE user_id = 4214\n"
            + "AND browser_cookie = 'aefde34ec'\n"
            + "AND ip_address = '10.10.0.2';");

    Approvals.verify(output);
  }

  @Test
  public void shouldAnonymizeCreateConnectorCorrectly() {
    final String output = anon.anonymize(
        "CREATE SOURCE CONNECTOR `jdbc-connector` WITH(\n"
            + "\"connector.class\"='io.confluent.connect.jdbc.JdbcSourceConnector',\n"
            + "\"connection.url\"='jdbc:postgresql://localhost:5432/my.db',\n"
            + "\"mode\"='bulk',\n"
            + "\"topic.prefix\"='jdbc-',\n"
            + "\"table.whitelist\"='users',\n"
            + "\"key\"='username');");

    Approvals.verify(output);
  }

  @Test
  public void shouldAnonymizeCreateTypeCorrectly() {
    // simple statement
    Assert.assertEquals("CREATE TYPE type AS INTEGER;",
        anon.anonymize("CREATE TYPE ADDRESS AS INTEGER;"));

    // more elaborate statement
    final String output = anon.anonymize(
        "CREATE TYPE ADDRESS AS STRUCT<number INTEGER, street VARCHAR, city VARCHAR>;");

    Approvals.verify(output);
  }

  @Test
  public void shouldAnonymizeAlterOptionCorrectly() {
    final String output = anon.anonymize(
        "ALTER STREAM my_stream ADD COLUMN c3 INT, ADD COLUMN c4 INT;");

    Approvals.verify(output);
  }

  @Test
  public void shouldAnonymizeInsertIntoCorrectly() {
    final String output = anon.anonymize(
        "INSERT INTO my_stream SELECT user_id, browser_cookie, ip_address\n"
            + "FROM another_stream\n"
            + "WHERE user_id = 4214\n"
            + "AND browser_cookie = 'aefde34ec'\n"
            + "AND ip_address = '10.10.0.2';");

    Approvals.verify(output);
  }

  @Test
  public void shouldAnonymizeInsertValuesCorrectly() {
    final String output = anon.anonymize(
        "INSERT INTO foo (ROWTIME, KEY_COL, COL_A) VALUES (1510923225000, 'key', 'A');");

    Approvals.verify(output);
  }

  @Test
  public void shouldAnonymizeDropStatementsCorrectly() {
    Assert.assertEquals("DROP STREAM IF EXISTS stream1 DELETE TOPIC;",
        anon.anonymize("DROP STREAM IF EXISTS my_stream DELETE TOPIC;"));
    Assert.assertEquals("DROP TABLE IF EXISTS table1 DELETE TOPIC;",
        anon.anonymize("DROP TABLE IF EXISTS my_table DELETE TOPIC;"));
    Assert.assertEquals("DROP CONNECTOR IF EXISTS connector;",
        anon.anonymize("DROP CONNECTOR IF EXISTS my_connector;"));
    Assert.assertEquals("DROP TYPE IF EXISTS type;",
        anon.anonymize("DROP TYPE IF EXISTS my_type;"));
  }

  @Test
  public void shouldAnonymizeUDFQueriesCorrectly() {
    final String output = anon.anonymize("CREATE STREAM OUTPUT AS SELECT ID, "
        + "REDUCE(numbers, 2, (s, x) => s + x) AS reduce FROM test;");

    Approvals.verify(output);
  }
}
