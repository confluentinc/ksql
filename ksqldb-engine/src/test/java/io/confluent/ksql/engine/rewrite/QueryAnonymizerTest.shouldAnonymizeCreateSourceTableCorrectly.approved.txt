CREATE SOURCE TABLE table1 (column1 VARCHAR, column2 DOUBLE, column3 DOUBLE) WITH (KAFKA_TOPIC='[string]', VALUE_FORMAT='[string]');